web/src/server/api/routers/utilities.ts

```diff
@@ -81,17 +82,102 @@ const isValidAndSecureUrl = async (urlString: string): Promise<boolean> => {
+  }
+};
+
+// Define a Zod schema for pre-signed S3 URL validation, based on https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html; 04.12.24
+const s3UrlSchema = z.object({
+  hostname: z
+    .string()
+    .refine((hostname) => hostname.endsWith(".amazonaws.com"), {
+      message: "Invalid hostname. Must be an S3 URL.",
+    }),
+  path: z.string(),
+  query: z.object({
+    "X-Amz-Algorithm": z.string().optional(),
+    "X-Amz-Credential": z.string().optional(),
+    "X-Amz-Date": z.string().optional(),
+    "X-Amz-Expires": z.string().optional(),
+    "X-Amz-SignedHeaders": z.string().optional(),
+    "X-Amz-Signature": z.string(),
+    "X-Amz-Security-Token": z.string().optional(),
+    "x-id": z.string().optional(), // Optional ID parameter, not documented in AWS docs as it is SDK specific, https://github.com/aws/aws-sdk-go-v2/blob/04e7aca073a0a7ed479aa37cad88a1cf58a979a1/service/s3/internal/customizations/presign_test.go#L35-L41
+  }),
+});
+
+/**
+ * Validate if a URL is a valid and live pre-signed S3 URL
+ * @param url The pre-signed S3 URL to validate
+ * @returns True if the URL is valid and reachable, false otherwise
+ */
+const isValidPresignedS3Url = async (url: string): Promise<boolean> => {
+  try {
+    const parsedUrl = new URL(url);
+
+    // Validate hostname and query parameters
+    const result = s3UrlSchema.safeParse({
+      hostname: parsedUrl.hostname,
+      path: parsedUrl.pathname,
+      query: Object.fromEntries(parsedUrl.searchParams.entries()),
+    });
+
+    if (!result.success) {
+      logger.info("Invalid pre-signed S3 URL:", result.error.message);
+      return false;
+    }
+
+    // Perform a HEAD request to check reachability
+    const response = await fetch(url, {
+      method: "HEAD",
+      signal: AbortSignal.timeout(5000),
+    });
+
+    // Status 200 indicates the URL is valid
+    if (response.ok) {
+      return true;
+    }
+
+    // Attempt a GET request as a fallback, as some pre-signed URLs are restricted to GET requests only
+    if (response.status === 403) {
+      logger.info(
+        "HEAD request returned 403, attempting GET for validation...",
+      );
+
+      const getResponse = await fetch(url, {
+        method: "GET",
+        signal: AbortSignal.timeout(5000),
+        headers: { Range: "bytes=0-1" }, // Fetch only the first byte, expected server response for valid pre-signed URLs is 206 Partial Content
+      });
+
+      return getResponse.ok; // 200 or 206 Partial Content indicates success
+    }
+
+    return false;
+  } catch (error) {
+    if (error instanceof z.ZodError) {
+      logger.info("URL Validation Error:", error.message);
+    } else if (error instanceof Error) {
+      logger.info("HEAD Request Error:", error.message);
+    } else {
+      logger.info("Unknown error:", error);
+    }
     return false;
   }
 };
 
 const isValidImageUrl = async (url: string): Promise<boolean> => {
   try {
+    if (await isValidPresignedS3Url(url)) {
```

Comment:
Could you add a 1-2 sentence comment on why S3 presigned URLs deserve special behaviour here? The head fetch + content type check seems sound and it's not obvious why they wouldn't work with a S3 presigned URL.

Reply:
Will do, yes 🙌